prov/verbs/src/windows/verbs_nd_ibv.c

```diff
@@ -0,0 +1,857 @@
+		errno = ret;
+		return NULL;
+	}
+	ofi_mutex_init(&channel_nd->q_lock);
+	channel_nd->num_pending = 0;
+	channel->context = context;
+	channel->fd = fd_signal_get(&channel_nd->q.signal);
+
+	return channel;
+}
+
+int ibv_destroy_comp_channel(struct ibv_comp_channel *channel)
+{
+	trace(__func__);
+	if (!channel) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct ibv_comp_channel_vrb_nd *channel_nd =
+		container_of(channel, struct ibv_comp_channel_vrb_nd, channel);
+
+	while (channel_nd->num_pending) {
+		int ret = dlistfd_wait_avail(&channel_nd->q, -1);
+		if (ret == 1) {
+			ofi_mutex_lock(&channel_nd->q_lock);
+			struct ibv_comp_event_vrb_nd *entry_nd =
+				container_of(channel_nd->q.list.next,
+					     struct ibv_comp_event_vrb_nd,
+					     entry);
+			dlistfd_remove(channel_nd->q.list.next, &channel_nd->q);
+			channel_nd->num_pending--;
+			free(entry_nd);
+			ofi_mutex_unlock(&channel_nd->q_lock);
+		}
+	}
+	dlistfd_head_free(&channel_nd->q);
+	ofi_mutex_destroy(&channel_nd->q_lock);
+	free(channel);
+	return FI_SUCCESS;
+}
+
+struct ibv_cq *ibv_create_cq(struct ibv_context *context, int cqe,
+			     void *cq_context, struct ibv_comp_channel *channel,
+			     int comp_vector)
+{
+	trace(__func__);
+	if (!context) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	struct adapter_t *adapter = get_adapter_by_context(context);
+	if (!adapter) {
+		errno = ENODEV;
+		return NULL;
+	}
+
+	struct ibv_cq *cq =
+		(struct ibv_cq *)calloc(1, sizeof(struct ibv_cq_vrb_nd));
+	if (!cq) {
+		errno = ENOMEM;
+		return NULL;
+	}
+	struct ibv_cq_vrb_nd *cq_nd =
+		container_of(cq, struct ibv_cq_vrb_nd, cq);
+
+	cq->context = context;
+	cq->channel = channel;
+	cq->cq_context = cq_context;
+	cq->cqe = cqe;
+
+	HRESULT hr = adapter->adapter->lpVtbl->CreateOverlappedFile(
+		adapter->adapter, &cq_nd->ov_file);
+	VRB_DBG(FI_LOG_CQ, "IND2Adapter::CreateOverlappedFile: hr=0x%08lx\n",
+		hr);
+	if (FAILED(hr)) {
+		errno = H2F(hr);
+		goto err1;
+	}
+
+	assert(cq_nd->ov_file && cq_nd->ov_file != INVALID_HANDLE_VALUE);
+
+	BindIoCompletionCallback(cq_nd->ov_file, vrb_nd_io_cb, 0);
+
+	hr = adapter->adapter->lpVtbl->CreateCompletionQueue(
+		adapter->adapter, &IID_IND2CompletionQueue, cq_nd->ov_file, cqe,
+		0, 0, &cq_nd->nd2cq);
+	VRB_DBG(FI_LOG_CQ, "IND2Adapter::CreateCompletionQueue: hr=0x%08lx\n",
+		hr);
+	if (FAILED(hr)) {
+		errno = H2F(hr);
+		goto err2;
+	}
+	assert(cq_nd->nd2cq);
+
+	return cq;
+err2:
+	CloseHandle(cq_nd->ov_file);
+err1:
+	free(cq);
+	return NULL;
+}
+
+int ibv_destroy_cq(struct ibv_cq *cq)
+{
+	trace(__func__);
+	if (!cq) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct ibv_cq_vrb_nd *cq_nd =
+		container_of(cq, struct ibv_cq_vrb_nd, cq);
+	HRESULT hr =
+		cq_nd->nd2cq->lpVtbl->CancelOverlappedRequests(cq_nd->nd2cq);
+	int ret = CloseHandle(cq_nd->ov_file);
+	assert(ret);
+	// Can't assert that the reference count is zero because releasing the
+	// associated QueuePair doesn't always release its CompletionQueues.
```

Comment:
What you can't read my mind? ;-)

I added this comment when I was working on making sure that all acquired resources were released. I think it's a bug in the Mellanox ND driver for my hardware.

I could update the comment to something like...
```
    // There appears to be a bug related to the reference counting for
    // ComplettionQueue objects in the Mellanox ND driver. CompletionQueue
    // reference counts are incremented when they are used in the creation
    // of a QueuePair, but are not always decremented when the QueuePair is
    // Released.  Ideally, we'd assert that that the reference count is zero
    // here, but we cannot because of that bug.
```
...or just drop it altogether.

Reply:
If there's a reproducible bug in the lower SW that we need to work around, I'd go with the comment you proposed.  Otherwise there's no way for a reader to figure out why the code is structured the way it is.